src/compiler/program.rs

```diff
@@ -40,53 +41,136 @@ fn create_compiler_host_worker() -> impl CompilerHost {
 }
 
 struct ProgramConcrete {
-    files: Vec<Rc<SourceFile>>,
+    files: Vec<Rc</*SourceFile*/ Node>>,
+    diagnostics_producing_type_checker: Option<TypeChecker>,
 }
 
 impl ProgramConcrete {
-    pub fn new(files: Vec<Rc<SourceFile>>) -> Self {
-        ProgramConcrete { files }
+    pub fn new(files: Vec<Rc<Node>>) -> Self {
+        ProgramConcrete {
+            files,
+            diagnostics_producing_type_checker: None,
+        }
+    }
+
+    fn get_diagnostics_producing_type_checker(&mut self) -> &TypeChecker {
+        // self.diagnostics_producing_type_checker
```

Comment:
This was two different ways that I tried to write this logic but where the borrow checker was unhappy (the pattern here is basically "return the already-cached value or initialize the cache and then return that newly created value")